### PR TITLE
Adding dummy_exclude table for GF180MCU DRC

### PR DESCRIPTION
**Pull request by FaragElsayed2**
## Adding dummy_exclude table for GF180MCU DRC

Fixes #43

### CODE REVIEW

**Review by atorkmabrains — APPROVED** [review]
Thanks @FaragElsayed2

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+## If size greater than 15000 um2 then two sides should not be greater than (80 um).
+logger.info('Executing rule DE.3')
+de3_ndmy_area = ndmy.with_area(15_000.um, nil).edges
+de3_l1 = de3_ndmy_area.with_length(80.001.um, nil)
+de3_l = de3_ndmy_area.join(de3_l1)
+de3_l.output('DE.3', 'DE.3 : Maximum NDMY size (um2): 15000 um2.
+              If size greater than 15000 um2 then two sides should not be greater than (80 um).')
+de3_l.forget
+de3_l1.forget
+de3_ndmy_area.forget
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:42]: nit: should we set those to `nil` (as suggested in https://www.klayout.de/doc/about/drc_ref_layer.html#h2-1260) to make sure they are not being used anymore.

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#================================================
+
+# Rule DE.1 is not a DRC check
+## Please refer to https://gf180mcu-pdk.readthedocs.io/en/latest/physical_verification/design_manual/drm_10_08.html#design-rules-for-dummy-exclude-layers-ndmy-and-pmndmy
+
+# Rule DE.2: Minimum NDMY or PMNDMY size (x or y dimension in um) is 0.8µm.
+logger.info('Executing rule DE.2')
+de2_l1 = ndmy.or(pmndmy).width(0.8.um, euclidian)
+de2_l1.output('DE.2', 'DE.2 : Minimum NDMY or PMNDMY size (x or y dimension in um) : 0.8µm')
+de2_l1.forget
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:30]: nit: should we set those to nil (as suggested in https://www.klayout.de/doc/about/drc_ref_layer.html#h2-1260) to make sure they are not being used anymore.

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################################
+
+#================================================
+#--------------DUMMY EXCLUDE LAYERS--------------
+#================================================
+
+# Rule DE.1 is not a DRC check
+## Please refer to https://gf180mcu-pdk.readthedocs.io/en/latest/physical_verification/design_manual/drm_10_08.html#design-rules-for-dummy-exclude-layers-ndmy-and-pmndmy
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:24]: should we still log some message?

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+              If size greater than 15000 um2 then two sides should not be greater than (80 um).')
+de3_l.forget
+de3_l1.forget
+de3_ndmy_area.forget
+
+# Rule DE.4: Minimum NDMY to NDMY space (Merge if space is less) is 20µm.
+logger.info('Executing rule DE.4')
+de4_l1 = ndmy.space(20.um, euclidian)
+de4_l1.output('DE.4', 'DE.4 : Minimum NDMY to NDMY space (Merge if space is less) : 20µm')
+de4_l1.forget
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:48]: nit: should we set those to nil (as suggested in https://www.klayout.de/doc/about/drc_ref_layer.html#h2-1260) to make sure they are not being used anymore.

**Review by proppy — APPROVED** [review]
just a few nit comments, but looks good :)

(it's great to see the CI and lint working!)

File: klayout/drc/testing/testcases/unit/dummy_exclude.svg [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+<text id="0x55914aa0dfb0" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1950.21 37.66) scale(2) scale(1 -1)">DE.3</text>
+<text id="0x55914aa6efd0" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1958.98 1087.92) scale(2) scale(1 -1)">DE.3</text>
+<text id="0x55914ab4f770" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1681.23 2140.12) scale(2) scale(1 -1)">DE.4</text>
+<text id="0x55914ab4f800" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1950.21 2140.12) scale(2) scale(1 -1)">DE.4</text>
+<text id="0x55914ab4f890" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1950.21 2509.93) scale(2) scale(1 -1)">DE.2</text>
+<text id="0x55914ab6ade0" class="l11t222" text-anchor="start" dominant-baseline="text-after-edge" transform="translate(-1467.55 2509.93) scale(2) scale(1 -1)">DE.2</text>
+<text id="0x55914ab6ae70" class="l58t0" text-anchor="start" dominant-baseline="central" transform="translate(-889.31 2620.75) scale(2) scale(1 -1)">2 ERRs</text>
```
Comment (proppy) [inline klayout/drc/testing/testcases/unit/dummy_exclude.svg:38]: the text mention 2 errors while there seems to be only 1 checked, is that expected?

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+logger.info('Executing rule DE.2')
+de2_l1 = ndmy.or(pmndmy).width(0.8.um, euclidian)
+de2_l1.output('DE.2', 'DE.2 : Minimum NDMY or PMNDMY size (x or y dimension in um) : 0.8µm')
+de2_l1.forget
+
+# Rule DE.3: Maximum NDMY size (um2) is 15000 um2.
+## If size greater than 15000 um2 then two sides should not be greater than (80 um).
+logger.info('Executing rule DE.3')
+de3_ndmy_area = ndmy.with_area(15_000.um, nil).edges
+de3_l1 = de3_ndmy_area.with_length(80.001.um, nil)
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:36]: shouldn't that be `de3_ndmy_area.with_length(nil, 80.001.um)`?

File: klayout/drc/rule_decks/dummy_exclude.drc [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+de3_l = de3_ndmy_area.join(de3_l1)
```
Comment (proppy) [inline klayout/drc/rule_decks/dummy_exclude.drc:37]: curious why we need to join here?

Comment (atorkmabrains) [reply klayout/drc/rule_decks/dummy_exclude.drc:37]: @proppy `join` works as `or` but much faster. It combines the outputs.

**Comment by atorkmabrains** [comment]
@proppy Could we merge this PR as we have addressed all your questions in other PRs?

File: klayout/drc/rule_decks/dummy_exclude.drc (hunk 2 above)
Comment (proppy) [reply klayout/drc/rule_decks/dummy_exclude.drc:30]: https://github.com/google/globalfoundries-pdk-libs-gf180mcu_fd_pv/issues/53

File: klayout/drc/rule_decks/dummy_exclude.drc (hunk 6 above)
Comment (proppy) [reply klayout/drc/rule_decks/dummy_exclude.drc:36]: https://github.com/google/globalfoundries-pdk-libs-gf180mcu_fd_pv/issues/54

File: klayout/drc/rule_decks/dummy_exclude.drc (hunk 4 above)
Comment (proppy) [reply klayout/drc/rule_decks/dummy_exclude.drc:48]: #53

File: klayout/drc/rule_decks/dummy_exclude.drc (hunk 1 above)
Comment (proppy) [reply klayout/drc/rule_decks/dummy_exclude.drc:42]: #53

File: klayout/drc/rule_decks/dummy_exclude.drc (hunk 3 above)
Comment (proppy) [reply klayout/drc/rule_decks/dummy_exclude.drc:24]: #55

File: klayout/drc/testing/testcases/unit/dummy_exclude.svg (hunk 5 above)
Comment (proppy) [reply klayout/drc/testing/testcases/unit/dummy_exclude.svg:38]: #56

**Comment by proppy** [comment]
> @proppy Could we merge this PR as we have addressed all your questions in other PRs?

Sure, I just needed to find some time to file issue for the items that were raised during the review, sorry for the delay!

**Pull request merged**